File: gfg/2024_7_july/c++/27_form_a_palindrome.cpp
// TIME COMPLEXITY
// O(n^2)

// SPACE COMPLEXITY
// O(n^2)

// CODE

class Solution{
  public:
    int LCS(int i, int j, string &s, string &p, vector<vector<int> > &dp){
        if(i == s.length() || j == p.length()) return 0;
        
        if(dp[i][j] != -1) return dp[i][j];
        
        if(s[i] == p[j]) return dp[i][j] = 1 + LCS(i+1, j+1, s,p,dp);
        return dp[i][j] = max(LCS(i, j+1, s,p,dp), LCS(i+1, j, s,p,dp));
    }
    int countMin(string str){
        string rev_str = str;
        reverse(rev_str.begin(), rev_str.end());
        if(rev_str == str) return 0;
        int n = str.length();
        vector<vector<int> > dp(n+1,vector<int>(n+1,-1)); 
        int lcs = LCS(0,0,str,rev_str,dp);
        return n-lcs;
    }
};

File: gfg/2024_7_july/python/27_form_a_palindrome.py
# CODE

class Solution:
    def countMin(self, str):
        n = len(str)
        a, b = str, str[::-1]
        LCS = [[0] * (n + 1) for _ in range(n + 1)]
        
        for i in range(1, n + 1):
            for j in range(1, n + 1):
                if a[i - 1] == b[j - 1]:
                    LCS[i][j] = 1 + LCS[i - 1][j - 1]
                else:
                    LCS[i][j] = max(LCS[i - 1][j], LCS[i][j - 1])
                    
        return n - LCS[n][n]